mat64/list_test.go

```diff
@@ -0,0 +1,706 @@
+				{5, 1, 1, 1},
+				{6, 6, 6, 11},
+				{6, 6, 11, 6},
+				{6, 11, 6, 6},
+				{11, 6, 6, 6},
+				{11, 11, 11, 6},
+				{11, 11, 6, 11},
+				{11, 6, 11, 11},
+				{6, 11, 11, 11},
+
+				{1, 1, 5, 5},
+				{1, 5, 1, 5},
+				{1, 5, 5, 1},
+				{5, 1, 1, 5},
+				{5, 1, 5, 1},
+				{5, 5, 1, 1},
+				{6, 6, 11, 11},
+				{6, 11, 6, 11},
+				{6, 11, 11, 6},
+				{11, 6, 6, 11},
+				{11, 6, 11, 6},
+				{11, 11, 6, 6},
+
+				{1, 1, 17, 11},
+				{1, 1, 11, 17},
+				{1, 11, 1, 17},
+				{1, 17, 1, 11},
+				{1, 11, 17, 1},
+				{1, 17, 11, 1},
+				{11, 1, 1, 17},
+				{17, 1, 1, 11},
+				{11, 1, 17, 1},
+				{17, 1, 11, 1},
+				{11, 17, 1, 1},
+				{17, 11, 1, 1},
+
+				{6, 6, 1, 11},
+				{6, 6, 11, 1},
+				{6, 11, 6, 1},
+				{6, 1, 6, 11},
+				{6, 11, 1, 6},
+				{6, 1, 11, 6},
+				{11, 6, 6, 1},
+				{1, 6, 6, 11},
+				{11, 6, 1, 6},
+				{1, 6, 11, 6},
+				{11, 1, 6, 6},
+				{1, 11, 6, 6},
+
+				{6, 6, 17, 1},
+				{6, 6, 1, 17},
+				{6, 1, 6, 17},
+				{6, 17, 6, 1},
+				{6, 1, 17, 6},
+				{6, 17, 1, 6},
+				{1, 6, 6, 17},
+				{17, 6, 6, 1},
+				{1, 6, 17, 6},
+				{17, 6, 1, 6},
+				{1, 17, 6, 6},
+				{17, 1, 6, 6},
+
+				{6, 6, 17, 11},
+				{6, 6, 11, 17},
+				{6, 11, 6, 17},
+				{6, 17, 6, 11},
+				{6, 11, 17, 6},
+				{6, 17, 11, 6},
+				{11, 6, 6, 17},
+				{17, 6, 6, 11},
+				{11, 6, 17, 6},
+				{17, 6, 11, 6},
+				{11, 17, 6, 6},
+				{17, 11, 6, 6},
+			} {
+				// Skip the test if the arguments are compile-time provably incorrect
+				// This is either because the matrix types are wrong at compile time
+				// or because the sizes are impossible
+				ok := legalTypes(aMat, bMat)
+				if !ok {
+					continue
+				}
+				if ok := dimsLegal(aMat, test.ar, test.ac); !ok {
+					continue
+				}
+				if ok := dimsLegal(bMat, test.br, test.bc); !ok {
+					continue
+				}
+				a := makeRandOf(aMat, test.ar, test.ac)
+				{
+					ar, ac := a.Dims()
+					if ar != test.ar || ac != test.ac {
+						panic("makeRandOf returns wrong size" + fmt.Sprintf("%T\n", aMat))
+						continue
+					}
+				}
+				b := makeRandOf(bMat, test.br, test.bc)
+				{
+					br, bc := b.Dims()
+					if br != test.br || bc != test.bc {
+						panic("makeRandOf returns wrong size" + fmt.Sprintf("%T\n", bMat))
+						continue
+					}
+				}
+
+				// Compute the true answer if the sizes are legal.
+				legal := legalSize(test.ar, test.ac, test.br, test.bc)
+				var truth Dense
+				if legal {
+					var aDense, bDense Dense
+					aDense.Clone(a)
+					bDense.Clone(b)
+					denseComparison(&truth, &aDense, &bDense)
+				}
+				// Call the method and check the result with all transpose variations.
+				// Do this by explicitly create a transpose, and then calling T().
+				for _, aTrans := range []bool{true, false} {
+					for _, bTrans := range []bool{true, false} {
+						aCopy := makeCopyOf(a)
+						bCopy := makeCopyOf(b)
+						// Set the proper function call
+						var f func(receiver, a, b Matrix)
+						if aTrans {
```

Comment:
It seems to me that this can be flattened into two conditions and a final call to method using a transient pair of Matrix values. This abolishes the need for 2 nested conditionals.